67_RayQueryGeometry/main.cpp

```diff
@@ -0,0 +1,1010 @@
+
+			smart_refctd_ptr<nbl::video::IGPUCommandBuffer> cmdbuf;
+			{
+				smart_refctd_ptr<nbl::video::IGPUCommandPool> cmdpool = m_device->createCommandPool(cQueue->getFamilyIndex(), IGPUCommandPool::CREATE_FLAGS::TRANSIENT_BIT);
+				if (!cmdpool->createCommandBuffers(IGPUCommandPool::BUFFER_LEVEL::PRIMARY, 1u, &cmdbuf))
+					return logFail("Failed to create one time Command Buffer!\n");
+			}
+
+			// create geometry objects
+			createGeometries(cmdbuf.get(), geometryCreator);
+
+			// create blas/tlas
+			createAccelerationStructures(cmdbuf.get());
+
+			// submit builds
+			{
+				auto completed = m_device->createSemaphore(0u);
+
+				std::array<IQueue::SSubmitInfo::SSemaphoreInfo, 1u> signals;
+				{
+					auto& signal = signals.front();
+					signal.value = 1;
+					signal.stageMask = bitflag(PIPELINE_STAGE_FLAGS::ALL_TRANSFER_BITS);
+					signal.semaphore = completed.get();
+				}
+
+				const IQueue::SSubmitInfo::SCommandBufferInfo commandBuffers[1] = { {
+					.cmdbuf = cmdbuf.get()
+				} };
+
+				const IQueue::SSubmitInfo infos[] =
+				{
+					{
+						.waitSemaphores = {},
+						.commandBuffers = commandBuffers,
+						.signalSemaphores = signals
+					}
+				};
+
+				if (cQueue->submit(infos) != IQueue::RESULT::SUCCESS)
+				{
+					m_logger->log("Failed to submit geometry transfer upload operations!", ILogger::ELL_ERROR);
+					return false;
+				}
+
+				const ISemaphore::SWaitInfo info[] =
+				{ {
+					.semaphore = completed.get(),
+					.value = 1
+				} };
+
+				m_device->blockForSemaphores(info);
+			}
+
+			// create pipelines
+			{
+				// shader
+				const std::string shaderPath = "render.comp.hlsl";
+				IAssetLoader::SAssetLoadParams lparams = {};
+				lparams.logger = m_logger.get();
+				lparams.workingDirectory = "";
+				auto bundle = m_assetMgr->getAsset(shaderPath, lparams);
+				if (bundle.getContents().empty() || bundle.getAssetType() != IAsset::ET_SHADER)
+				{
+					m_logger->log("Shader %s not found!", ILogger::ELL_ERROR, shaderPath);
+					exit(-1);
+				}
+
+				const auto assets = bundle.getContents();
+				assert(assets.size() == 1);
+				smart_refctd_ptr<ICPUShader> shaderSrc = IAsset::castDown<ICPUShader>(assets[0]);
+				shaderSrc->setShaderStage(IShader::E_SHADER_STAGE::ESS_COMPUTE);
+				auto shader = m_device->createShader(shaderSrc.get());
+
+				// descriptors
+				IGPUDescriptorSetLayout::SBinding bindings[] = {
+					{
+						.binding = 0,
+						.type = asset::IDescriptor::E_TYPE::ET_ACCELERATION_STRUCTURE,
+						.createFlags = IGPUDescriptorSetLayout::SBinding::E_CREATE_FLAGS::ECF_NONE,
+						.stageFlags = asset::IShader::E_SHADER_STAGE::ESS_COMPUTE,
+						.count = 1,
+					},
+					{
+						.binding = 1,
+						.type = asset::IDescriptor::E_TYPE::ET_UNIFORM_BUFFER,
+						.createFlags = IGPUDescriptorSetLayout::SBinding::E_CREATE_FLAGS::ECF_NONE,
+						.stageFlags = asset::IShader::E_SHADER_STAGE::ESS_COMPUTE,
+						.count = 1,
+					},
```

Comment:
if your uniforms can fit in push constants, don't complicate your life with UBOs

Reply:
still a TODO